### PR TITLE
feat: [SPG-155]: Module Configuration initial commit

### PR DESCRIPTION
**Pull request by avinashmadhwani02**
#### Screenshots

NA

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier: `fix prettier`
</details>

## [Contributor license agreement](https://github.com/harness/harness-core-ui/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 182 Issues 
* Code issue 182 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/e864badd0b95ab81c060b15b85f15701/pull-request/11021)

File: src/modules/10-common/navigation/ModuleConfigurationScreen/useGetContentfulModules.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+import { useEffect, useState } from 'react'
+import type { Entry, Asset } from 'contentful'
+import Contentful from '@common/Contentful'
+import type { NavModuleName } from '@common/hooks/useNavModuleInfo'
+
+const CONTENT_TYPE = 'module'
+
+export interface LottieContent {
+  data: object
```
Comment (abhinavrastogi-harness) [inline src/modules/10-common/navigation/ModuleConfigurationScreen/useGetContentfulModules.ts:16]: dont use `object` as type.
use `Record<>` instead

File: src/modules/10-common/navigation/ModuleConfigurationScreen/ModuleDetailsSection/ModuleCarousel.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+import React from 'react'
+import Lottie from 'react-lottie-player'
+import { Icon } from '@harness/icons'
+import { Container, Layout, Text, FontVariation, Color, Carousel } from '@harness/uicore'
+import useNavModuleInfo, { NavModuleName } from '@common/hooks/useNavModuleInfo'
+import { MassagedModuleData, ModuleContentType } from '../useGetContentfulModules'
+import CarouselImageAndDescription from '../CarousellmageAndDescription/CarousellmageAndDescription'
+import LottieComponent from '../LottieComponent/LottieComponent'
+import defaultLottie from './default_lottie.json'
```
Comment (abhinavrastogi-harness) [inline src/modules/10-common/navigation/ModuleConfigurationScreen/ModuleDetailsSection/ModuleCarousel.tsx:16]: this is a huuuge file. make it a dynamic import and ensure it's loaded only when needed.

File: src/modules/10-common/navigation/MainNav/MainNav.module.scss [hunk 3]
```diff
@@ -18,10 +25,34 @@
...
     list-style: none;
     padding: 0;
     margin: 0;
+
+    .modulesContainer {
+      overflow-y: scroll;
+      background: var(--modules-container-bg);
+      border-bottom: 1px solid var(--modules-container-border);
+      border-top: 1px solid var(--modules-container-border);
+      &::-webkit-scrollbar {
```
[the remark this answers is not in the document]
Comment (mayankVermaHarness) [reply src/modules/10-common/navigation/MainNav/MainNav.module.scss:51]: Do we use something like PostCSS for this?

File: src/modules/10-common/navigation/MainNav/MainNav.tsx [hunk 4]
```diff
@@ -19,29 +20,71 @@ import paths from '@common/RouteDefinitions'
...
+  STONavItem
+} from './ModuleLinks'
 import css from './MainNav.module.scss'
 
 const commonLinkProps: Partial<NavLinkProps> = {
   activeClassName: css.active,
   className: cx(css.navLink)
 }
 
+const maxNumOfModulesToShow = 3
```
Comment (mayankVermaHarness) [inline src/modules/10-common/navigation/MainNav/MainNav.tsx:55]: Can we have this capitalized like: `MAX_NUM_OF_MODULES_TO_SHOW` ?

File: src/modules/10-common/navigation/ModuleConfigurationScreen/CarousellmageAndDescription/__tests__/CarouselImageAndDescription.test.tsx [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  test('render', () => {
+    const { container, getByText } = render(
+      <TestWrapper>
+        <CarouselImageAndDescription
+          activeModule={ModuleName.CD}
+          primaryText="primary text"
+          secondaryText="secondory text"
+          // eslint-disable-next-line
+          // @ts-ignore
+          image={{
```
Comment (mayankVermaHarness) [inline src/modules/10-common/navigation/ModuleConfigurationScreen/CarousellmageAndDescription/__tests__/CarouselImageAndDescription.test.tsx:24]: Can we define this `object` outside the `render` for better readability?

File: src/modules/10-common/navigation/ModuleConfigurationScreen/LottieComponent/LottieComponent.tsx [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+}
+
+const LottieComponent: React.FC<LottieComponentProps> = ({ json: asset }) => {
+  const [lottieJson, setLottieJson] = useState<object | undefined>(undefined)
+  const [loading, setLoading] = useState(false)
+
+  useEffect(() => {
+    if (!lottieJson) {
+      setLoading(true)
+      fetch(`https://${asset.fields.file.url}`)
```
Comment (mayankVermaHarness) [inline src/modules/10-common/navigation/ModuleConfigurationScreen/LottieComponent/LottieComponent.tsx:26]: what if someone uses this without `https`?

Can we use `window.location.protocol` or simply start with `//` like this: 

`fetch( "//${asset.fields.file.url}" )`

File: src/modules/10-common/navigation/ModuleList/NavModule/NavModule.tsx [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+const navModuleToClassMap: Record<NavModuleName, string> = {
+  [ModuleName.CD]: css.cd,
+  [ModuleName.CI]: css.ci,
+  [ModuleName.CV]: css.srm,
+  [ModuleName.CF]: css.ff,
+  [ModuleName.CE]: css.ccm,
+  [ModuleName.STO]: css.sto,
+  [ModuleName.CHAOS]: css.chaos,
+  [ModuleName.SCM]: css.ff // change this
```
Comment (mayankVermaHarness) [inline src/modules/10-common/navigation/ModuleList/NavModule/NavModule.tsx:38]: Change this to SCM ?

File: src/modules/10-common/navigation/ModuleList/__tests__/ModuleList.test.tsx [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    expect(container.querySelector('[data-icon="customize"]')).not.toBeNull()
+    expect(getByText('common.moduleList.title')).toBeDefined()
+    expect(getByTestId('grouplistContainer').children.length).toEqual(0)
+    expect(container).toMatchSnapshot()
+  })
+
+  test('should render correctly with modules enabled and disabled', () => {
+    const { queryByText } = render(
+      <TestWrapper
+        path="/account/:accountId"
```
Comment (mayankVermaHarness) [inline src/modules/10-common/navigation/ModuleList/__tests__/ModuleList.test.tsx:37]: Can use `routes.To<Path>`?

File: src/modules/10-common/Contentful.tsx [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+import { createClient, ContentfulClientApi } from 'contentful'
+
+export enum ContentfulEnvironment {
+  PRODUCTION = 'master'
+}
+
+class Contentful {
+  private client: ContentfulClientApi | undefined = undefined
```
[the remark this answers is not in the document]
Comment (avinashmadhwani02) [reply src/modules/10-common/Contentful.tsx:15]: As discussed, for now we can go ahead with this. We will add the support if we need it in the future

File: src/modules/10-common/navigation/MainNav/MainNav.module.scss [hunk 10]
```diff
@@ -18,10 +25,34 @@
...
+        width: 0px;
```
[the remark this answers is not in the document]
Comment (avinashmadhwani02) [reply src/modules/10-common/navigation/MainNav/MainNav.module.scss:52]: Removed for now

File: src/modules/10-common/navigation/MainNav/MainNav.module.scss (hunk 3 above)
[the remark this answers is not in the document]
Comment (avinashmadhwani02) [reply src/modules/10-common/navigation/MainNav/MainNav.module.scss:51]: Removed this for now

File: src/modules/10-common/navigation/MainNav/MainNav.tsx [hunk 11]
```diff
@@ -19,29 +20,67 @@ import paths from '@common/RouteDefinitions'
...
   const { currentUserInfo: user } = useAppStore()
+  const { module } = useModuleInfo()
+  const { licenseInformation } = useLicenseStore()
+  const { preference: modulesPreferenceData, setPreference: setModuleConfigPreference } =
+    usePreferenceStore<ModulesPreferenceStoreData>(PreferenceScope.USER, MODULES_CONFIG_PREFERENCE_STORE_KEY)
+
+  const { selectedModules = [] } = modulesPreferenceData || {}
+  const modulesToShow = [...selectedModules]
+  // if current module is not selecting in the modules config, add it temporarily
+  if (module && !modulesToShow.includes(moduleToModuleNameMapping[module] as NavModuleName)) {
```
[the remark this answers is not in the document]
Comment (avinashmadhwani02) [reply src/modules/10-common/navigation/MainNav/MainNav.tsx]: Removed this logic. we don't need it

File: src/modules/10-common/navigation/ModuleConfigurationScreen/ModuleConfigurationScreen.module.scss [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+ * Use of this source code is governed by the PolyForm Shield 1.0.0 license
+ * that can be found in the licenses directory at the root of this repository, also available at
+ * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
+ */
+
+.container {
+  position: fixed;
+  top: 0;
+  left: 88px;
+  height: 100% !important;
```
[the remark this answers is not in the document]
Comment (avinashmadhwani02) [reply src/modules/10-common/navigation/ModuleConfigurationScreen/ModuleConfigurationScreen.module.scss:12]: Position is fixed. Will it not cover the entire screen always?

File: src/modules/10-common/navigation/ModuleConfigurationScreen/ModuleConfigurationScreen.tsx [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+          <String stringID="common.moduleConfig.selectModules" />
+        </Text>
+        <Text inline color={Color.PRIMARY_5} className={css.blueText} margin={{ left: 'small', right: 'small' }}>
+          <String stringID="common.moduleConfig.your" />
+        </Text>
+        <Text inline color={Color.WHITE} font={{ variation: FontVariation.H2 }}>
+          <String stringID="common.moduleConfig.navigation" />
+        </Text>
+      </Text>
+      {/* <Text className={css.defaultSettingsTextContainer}>
```
[the remark this answers is not in the document]
Comment (avinashmadhwani02) [reply src/modules/10-common/navigation/ModuleConfigurationScreen/ModuleConfigurationScreen.tsx]: uncommented

File: src/modules/10-common/hooks/__tests__/useNavModuleInfo.test.tsx [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+      initialProps: { path: '/account/my_account_id/cd/orgs/my_org/projects/my_project' }
+    })
+
+    expect(result.current.length).toBe(0)
+  })
+
+  test('test cd', () => {
+    const { result } = renderHook(() => useNavModuleInfo([ModuleName.CD]), {
+      wrapper: TestWrapper,
+      initialProps: { path: '/account/my_account_id/cd/orgs/my_org/projects/my_project' }
```
[the remark this answers is not in the document]
Comment (avinashmadhwani02) [reply src/modules/10-common/hooks/__tests__/useNavModuleInfo.test.tsx:26]: we don't need the path, removed it

File: src/modules/10-common/navigation/ModuleList/__tests__/ModuleList.test.tsx (hunk 8 above)
Comment (avinashmadhwani02) [reply src/modules/10-common/navigation/ModuleList/__tests__/ModuleList.test.tsx:37]: this path was not being used. removed it

**Review by abhinavrastogi-harness — APPROVED** [review]

File: src/modules/10-common/navigation/ModuleConfigurationScreen/ModuleSortableList/DraggableModuleItem/DraggableModuleItem.module.scss [hunk 15]
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright 2022 Harness Inc. All rights reserved.
+ * Use of this source code is governed by the PolyForm Shield 1.0.0 license
+ * that can be found in the licenses directory at the root of this repository, also available at
+ * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
+ */
+
+// .dragIcon {
```
Comment (mayankVermaHarness) [inline src/modules/10-common/navigation/ModuleConfigurationScreen/ModuleSortableList/DraggableModuleItem/DraggableModuleItem.module.scss:8]: Remove unused code?

**Review by mayankVermaHarness — APPROVED** [review]

**Review by mtran7 — APPROVED** [review]

**Pull request merged**